keps/sig-cluster-lifecycle/0035-20190128-addons-via-operators.md

```diff
@@ -0,0 +1,233 @@
+* Operators are able to observe other CRDs to perform basic sequencing
+* Addon manifests are able express an operator minimum version requirement, so
+  that an addon with new requirements can require that the operator be updated
+  first
+* Airgapped operation should be possible by combining a registry mirror and
+  storage of the underlying manifest in the cluster itself.
+
+
+An example can make this easier to understand, here is what a CRD instance for
+kube-proxy might look like:
+
+```yaml
+apiVersion: addons.sigs.k8s.io/v1alpha1
+kind: KubeProxy
+metadata:
+  name: default
+  namespace: kube-system
+spec:
+  clusterCidr: 100.64.0.0/10
+  version: 1.14.4
+status:
+  healthy: true
+```
+
+This particular manifest is pinned to `version: 1.14.4`.  We could also
+subscribe to a stream of updates with a field like `channel: stable`.
+
+### Risks and Mitigations
+
+This will involve running a large number of new controllers.  This will require
+more resources; we can mitigate this by combining them into a single binary
+(similar to kube-controller-manager).
+
+Automatically updating addons could result in new SPOFs, we can mitigate this
+through mirroring (including support for air-gapped mirrors).
+
+Providing a good set of addons could result in a monoculture where mistakes
+affect most/all kubernetes clusters (even if we don’t mandate adoption, if we
+succeed we hope for widespread adoption).  We can continue with our strategies
+that we use for core components such as kube-apiserver: primarily we must keep
+the notion of stable vs less-stable releases, to stagger the risk of a bad
+rollout.  We must also consider this a trade-off against the risk that without
+coordination each piece of tooling must reinvent the wheel; we expect more
+mistakes (even measured per cluster) in that scenario.
+
+Test and release may become more complicated because of fragmentation across
+repos.  Mitigation: be disciplined about versioning of operators and addons and
+encourage installation tooling to pin to a particular version of both for a
+particular release.  We need to set up automated builds (with CI) for rapid
+releases so installation tooling is not blocked waiting for operator releases.
+We need to set up a mechanism so that addons can be updated without requiring an
+operator update.  With this, if tooling is able to pin to particular addon
+versions, that should be at parity with the "embedded manifest" approach that is
+widely used currently.  (We hope to enable usage that is less lock-step, but
+that itself will likely require new approaches for testing and release)
+
+## Success Criteria
+
+We will succeed if addon operators are:
+
+* Used: addon operators are adopted by the majority of cluster installation
+tooling
+* Useful: users are generally satisfied with the functionality of addon
+operators and are not trying to work around them, or making lots of proposals /
+PRs to extend them
```

Comment:
I think this is a great success criteria and the one I'm skeptical about in the current scope.

Would we do an ingress controller operator?  Which ingress controllers would it support?  What options on those ingress controllers?

Reply:
If these all have operators, I was imagining there would be an nginx-ingress operator, and a distinct gclb-ingress operator, aws-alb operator etc.  Ideally component projects end up owning their operators.

I was not imagining we would try here to define a generic ingress operator that can install arbitrary ingress controllers  - it doesn't feel very compatible with the idea of federating responsibility & of self-determination.  It also feels like that would be a sig-network project.